### PR TITLE
Fix filter pushdown for extension plans

### LINKED ISSUE #5424: Filter Pushdown no longer works for extension plans
**Describe the bug**

The filter pushdown optimizer will not push down predicates through extension plans even if the extension plan allows it.

**To Reproduce**

```
    #[derive(Debug)]
    struct NoopPlan {
        input: Vec<LogicalPlan>,
        schema: DFSchemaRef,
    }

    impl UserDefinedLogicalNode for NoopPlan {
        fn as_any(&self) -> &dyn Any {
            self
        }

        fn inputs(&self) -> Vec<&LogicalPlan> {
            self.input.iter().collect()
        }

        fn schema(&self) -> &DFSchemaRef {
            &self.schema
        }

        fn expressions(&self) -> Vec<Expr> {
            self.input
                .iter()
                .flat_map(|child| child.expressions())
                .collect()
        }

        fn prevent_predicate_push_down_columns(&self) -> HashSet<String> {
            HashSet::from_iter(vec!["c".to_string()])
        }

        fn fmt_for_explain(&self, f: &mut Formatter) -> std::fmt::Result {
            write!(f, "NoopPlan")
        }

        fn from_template(
            &self,
            _exprs: &[Expr],
            inputs: &[LogicalPlan],
        ) -> Arc<dyn UserDefinedLogicalNode> {
            Arc::new(Self {
                input: inputs.iter().cloned().collect(),
                schema: self.schema.clone(),
            })
        }
    }

    #[test]
    fn user_defined_plan() -> Result<()> {
        let table_scan = test_table_scan()?;

        let custom_plan = LogicalPlan::Extension(Extension {
            node: Arc::new(NoopPlan {
                input: vec![table_scan.clone()],
                schema: table_scan.schema().clone(),
            }),
        });
        let plan = LogicalPlanBuilder::from(custom_plan)
            .filter(col("a").eq(lit(1i64)))?
            .build()?;

        // Push filter below NoopPlan
        let expected = "\
            NoopPlan\
            \n  Filter: test.a = Int64(1)\
            \n    TableScan: test";
        assert_optimized_plan_eq(&plan, expected)?;
   }
```

**Expected behavior**

If a predicate does not contain any columns in `UserDefinedLogicalNode::prevent_predicate_push_down_columns` it should get pushed down

**Additional context**

### PR DESCRIPTION
**Pull request by thinkharderdev**
# Which issue does this PR close?

Closes #5424 

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

Yes

# Are there any user-facing changes?

Predicates will be pushed down through extension plan nodes

No

### CODE REVIEW

File: datafusion/optimizer/src/push_down_filter.rs [hunk 1]
```diff
@@ -1029,6 +1074,131 @@ mod tests {
...
+            node: Arc::new(NoopPlan {
+                input: vec![table_scan.clone()],
+                schema: table_scan.schema().clone(),
+            }),
+        });
+        let plan = LogicalPlanBuilder::from(custom_plan)
+            .filter(col("a").eq(lit(1i64)).and(col("c").eq(lit(2i64))))?
+            .build()?;
+
+        // Push only predicate on `a` below NoopPlan
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_filter.rs:1154]: 👍

**Review by alamb — APPROVED** [review]
Thank you for the contribution @thinkharderdev  - the code makes sense to me and I reviewed the test coverage carefully. Thank you

File: datafusion/optimizer/src/push_down_filter.rs [hunk 2]
```diff
@@ -735,6 +735,48 @@ impl OptimizerRule for PushDownFilter {
...
+                            Ok(LogicalPlan::Filter(Filter::try_new(
+                                predicate.clone(),
+                                Arc::new(child.clone()),
+                            )?))
+                        })
+                        .collect::<Result<Vec<_>>>()?,
+                    None => extension_plan.node.inputs().into_iter().cloned().collect(),
+                };
+                let new_extension =
+                    from_plan(&filter.input, &filter.input.expressions(), &new_children)?;
```
Comment (jackwener) [inline datafusion/optimizer/src/push_down_filter.rs:770]:
```suggestion
                // extension with new inputs.
                let new_extension = child_plan.with_new_inputs(&new_children)?;
```

**Review by jackwener — APPROVED** [review]
A nice job, thanks @thinkharderdev

**Pull request merged**